core/src/software/aws/toolkits/core/credentials/sso/SsoAccessTokenProvider.kt

```diff
@@ -0,0 +1,135 @@
+    private val onPendingToken: SsoLoginCallback,
+    private val cache: SsoCache,
+    private val client: SsoOidcClient,
+    private val clock: Clock = Clock.systemUTC()
+) {
+    suspend fun accessToken(): AccessToken {
+        cache.loadAccessToken(ssoUrl)?.let {
+            return it
+        }
+
+        val token = pollForToken()
+
+        cache.saveAccessToken(ssoUrl, token)
+
+        return token
+    }
+
+    private fun registerClient(): ClientRegistration {
+        cache.loadClientRegistration(ssoRegion)?.let {
+            return it
+        }
+
+        // Based on botocore: https://github.com/boto/botocore/blob/5dc8ee27415dc97cfff75b5bcfa66d410424e665/botocore/utils.py#L1769
+        val registerResponse = client.registerClient {
+            it.clientType(CLIENT_REGISTRATION_TYPE)
+            it.clientName("aws-toolkit-jetbrains-${Instant.now(clock)}")
+        }
+
+        val registeredClient = ClientRegistration(
+            registerResponse.clientId(),
+            registerResponse.clientSecret(),
+            Instant.ofEpochSecond(registerResponse.clientSecretExpiresAt())
+        )
+
+        cache.saveClientRegistration(ssoRegion, registeredClient)
+
+        return registeredClient
+    }
+
+    private fun authorizeClient(clientId: ClientRegistration): Authorization {
+        // Should not be cached, only good for 1 token and short lived
+        val authorizationResponse = try {
+            client.startDeviceAuthorization {
+                it.startUrl(ssoUrl)
+                it.clientId(clientId.clientId)
+                it.clientSecret(clientId.clientSecret)
+            }
+        } catch (e: InvalidClientException) {
+            cache.invalidateClientRegistration(ssoRegion)
+            throw e
+        }
+
+        return Authorization(
+            authorizationResponse.deviceCode(),
+            authorizationResponse.userCode(),
+            authorizationResponse.verificationUri(),
+            authorizationResponse.verificationUriComplete(),
+            Instant.now(clock).plusSeconds(authorizationResponse.expiresIn().toLong()),
+            authorizationResponse.interval()?.toLong()
+                ?: DEFAULT_INTERVAL
+        )
+    }
+
+    private suspend fun pollForToken(): AccessToken {
+        val registration = registerClient()
+        val authorization = authorizeClient(registration)
+
+        onPendingToken.tokenPending(authorization)
+
+        var backOffTime = Duration.ofSeconds(authorization.pollInterval)
+
+        while (true) {
+            try {
+                val tokenResponse = client.createToken {
+                    it.clientId(registration.clientId)
+                    it.clientSecret(registration.clientSecret)
+                    it.grantType(GRANT_TYPE)
+                    it.deviceCode(authorization.deviceCode)
+                }
+
+                val expirationTime = Instant.now(clock).plusSeconds(tokenResponse.expiresIn().toLong())
```

Comment:
does it make sense to base it off server time instead? do we need to think about local clock drift or is the expiration time flexible for a few minutes?

Reply:
They don't return it, its a flaw in the OIDC spec

Reply:
https://tools.ietf.org/html/rfc8628#section-3.2

Reply:
We actually ignore the value all together, and just wait for the service to tell us to abort.